core/vault.go

```diff
@@ -0,0 +1,416 @@
+
+	res, err := vault.Do("GET", fmt.Sprintf("/v1/secret/%s", path), nil)
+	if err != nil {
+		return nil, exists, err
+	}
+	if res.StatusCode == 404 {
+		return nil, exists, nil
+	}
+	if res.StatusCode != 200 && res.StatusCode != 204 {
+		return nil, exists, fmt.Errorf("API %s", res.Status)
+	}
+
+	exists = true
+	b, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, exists, err
+	}
+
+	var raw map[string]interface{}
+	if err = json.Unmarshal(b, &raw); err != nil {
+		return nil, exists, err
+	}
+
+	if x, ok := raw["data"]; ok {
+		return x.(map[string]interface{}), exists, nil
+	}
+
+	return nil, exists, fmt.Errorf("Malformed response from Vault")
+}
+
+func (vault *Vault) Put(path string, data interface{}) error {
+	res, err := vault.Do("POST", fmt.Sprintf("/v1/secret/%s", path), data)
+	if err != nil {
+		return err
+	}
+	if res.StatusCode != 200 && res.StatusCode != 204 {
+		return fmt.Errorf("API %s", res.Status)
+	}
+	return nil
+}
+
+func (vault *Vault) ReadConfig(store string, master string) (VaultCreds, error) {
+	key := sha256.Sum256([]byte(master))
+	log.Debugf("reading credentials files from %s", store)
+	b, err := ioutil.ReadFile(store)
+	if err != nil {
+		log.Errorf("failed to read vault credentials from %s: %s", store, err)
+		return VaultCreds{}, err
+	}
+	creds := VaultCreds{}
+	err = json.Unmarshal(b, &creds)
+	if err != nil {
+		log.Errorf("failed to parse vault credentials from %s: %s", store, err)
+		return VaultCreds{}, err
+	}
+
+	plainSeal, err := vault.decrypt(key[:], []byte(creds.SealKey))
+	if err != nil {
+		log.Errorf("Failed to encrypt sealkey for longterm storage: %s", err)
+		return VaultCreds{}, err
+	}
+
+	plainRoot, err := vault.decrypt(key[:], []byte(creds.RootToken))
+	if err != nil {
+		log.Errorf("Failed to encrypt root token for longterm storage: %s", err)
+		return VaultCreds{}, err
+	}
+
+	plainType, err := vault.decrypt(key[:], []byte(creds.EncryptionType))
+	if err != nil {
+		log.Errorf("Failed to encrypt type for longterm storage: %s", err)
+		return VaultCreds{}, err
+	}
+
+	plainCreds := VaultCreds{
+		SealKey:        plainSeal,
+		RootToken:      plainRoot,
+		EncryptionType: plainType,
+	}
+
+	return plainCreds, err
+}
+
+func (vault *Vault) WriteConfig(store string, master string, creds VaultCreds) error {
+
+	key := sha256.Sum256([]byte(master))
+	encSeal, err := vault.encrypt(key[:], []byte(creds.SealKey))
+	if err != nil {
+		log.Errorf("Failed to encrypt sealkey for longterm storage: %s", err)
+		return err
+	}
+
+	encRoot, err := vault.encrypt(key[:], []byte(creds.RootToken))
+	if err != nil {
+		log.Errorf("Failed to encrypt root token for longterm storage: %s", err)
+		return err
+	}
+
+	encType, err := vault.encrypt(key[:], []byte(creds.EncryptionType))
+	if err != nil {
+		log.Errorf("Failed to encrypt type for longterm storage: %s", err)
+		return err
+	}
+
+	encryptedCreds := VaultCreds{
+		SealKey:        encSeal,
+		RootToken:      encRoot,
+		EncryptionType: encType,
```

Comment:
This should move out of here and into core config.  It's not sensitive, and needs to be modifiable by the admin.